sdk/pinocchio/src/sysvars/slot_hashes.rs

```diff
@@ -0,0 +1,420 @@
+    let data_len = buffer_len - NUM_ENTRIES_SIZE;
+    if data_len % ENTRY_SIZE != 0 {
+        return Err(ProgramError::InvalidArgument);
+    }
+
+    let max_entries = data_len / ENTRY_SIZE;
+    if max_entries > MAX_ENTRIES {
+        return Err(ProgramError::InvalidArgument);
+    }
+
+    // If declared entry count provided, validate it
+    if let Some(declared) = declared_entries {
+        if declared > MAX_ENTRIES {
+            return Err(ProgramError::InvalidArgument);
+        }
+        if declared > max_entries {
+            return Err(ProgramError::InvalidArgument);
+        }
+        return Ok(declared);
+    }
+
+    Ok(max_entries)
+}
+
+/// Validates SlotHashes data format and returns the entry count.
+fn parse_and_validate_data(data: &[u8]) -> Result<usize, ProgramError> {
+    // Need at least the 8-byte length prefix.
+    if data.len() < NUM_ENTRIES_SIZE {
+        return Err(ProgramError::AccountDataTooSmall);
+    }
+
+    let num_entries = unsafe { read_entry_count_from_bytes_unchecked(data) };
+
+    validate_slothashes_constraints(data.len(), Some(num_entries))
+}
+
+impl SlotHashEntry {
+    /// Returns the slot number as a u64.
+    #[inline(always)]
+    pub fn slot(&self) -> Slot {
+        u64::from_le_bytes(self.slot_le)
+    }
+}
+
+impl<T: Deref<Target = [u8]>> SlotHashes<T> {
+    /// Validates that a buffer is properly sized for SlotHashes data.
+    ///
+    /// Checks that the buffer length is 8 + (N * 40) for some N ≤ 512.
+    #[inline]
+    fn validate_buffer_size(buffer_len: usize) -> Result<(), ProgramError> {
+        validate_slothashes_constraints(buffer_len, None)?;
+        Ok(())
+    }
+
+    /// Creates a `SlotHashes` instance from arbitrary data with full validation.
+    ///
+    /// This constructor performs comprehensive validation of the data format
+    /// including length prefix, entry count bounds, and buffer size requirements.
+    /// Does not validate that entries are sorted in descending order.
+    pub fn new(data: T) -> Result<Self, ProgramError> {
+        let num_entries = parse_and_validate_data(&data)?;
+        Ok(unsafe { Self::new_unchecked(data, num_entries) })
+    }
+
+    /// Creates a `SlotHashes` instance directly from a data container and entry count.
+    /// Important: provide a valid len. Whether or not len is assumed to be
+    /// the constant 20_488 (512 entries) is up to caller.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because it does not check the validity of the data or count.
+    /// The caller must ensure:
+    /// 1. The underlying byte slice in `data` represents valid SlotHashes data
+    ///    (length prefix + entries, where entries are sorted in descending order by slot).
+    /// 2. `len` is the correct number of entries (≤ MAX_ENTRIES), matching the prefix.
+    /// 3. The data slice contains at least `NUM_ENTRIES_SIZE + len * ENTRY_SIZE` bytes.
+    /// 4. Alignment is correct for SlotHashEntry access.
```

Comment:
I think `4` is alway guaranteed since the alignment is `1` right?

Reply:
Thanks - old comment from before the alignment change and assertion